internal/gcs/process.go

```diff
@@ -0,0 +1,173 @@
+package gcs
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/logfields"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	hrNotFound = 0x80070490
+)
+
+type Process struct {
+	gc                    *GuestConnection
+	cid                   string
+	id                    uint32
+	waitCall              *rpc
+	waitResp              containerWaitForProcessResponse
+	stdin, stdout, stderr *ioChannel
+	stdinCloseWriteOnce   sync.Once
+	stdinCloseWriteErr    error
+}
+
+type baseProcessParams struct {
+	CreateStdInPipe, CreateStdOutPipe, CreateStdErrPipe bool
+}
+
+func (gc *GuestConnection) exec(ctx context.Context, cid string, params interface{}) (cow.Process, error) {
+	b, err := json.Marshal(params)
+	if err != nil {
+		return nil, err
+	}
+	var bp baseProcessParams
+	err = json.Unmarshal(b, &bp)
+	if err != nil {
+		return nil, err
+	}
+
+	req := containerExecuteProcess{
+		requestBase: makeRequest(cid),
+		Settings: executeProcessSettings{
+			ProcessParameters:       anyInString{params},
+			VsockStdioRelaySettings: &executeProcessVsockStdioRelaySettings{},
+		},
+	}
+
+	p := &Process{gc: gc, cid: cid}
+	if bp.CreateStdInPipe {
+		p.stdin, req.Settings.VsockStdioRelaySettings.StdIn, err = gc.newIoChannel()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if bp.CreateStdOutPipe {
+		p.stdout, req.Settings.VsockStdioRelaySettings.StdOut, err = gc.newIoChannel()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if bp.CreateStdErrPipe {
+		p.stderr, req.Settings.VsockStdioRelaySettings.StdErr, err = gc.newIoChannel()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var resp containerExecuteProcessResponse
+	err = gc.brdg.RPC(ctx, rpcExecuteProcess, &req, &resp, false)
+	if err != nil {
+		return nil, err
+	}
+	p.id = resp.ProcessID
+	// Start a wait message.
+	waitReq := containerWaitForProcess{
+		requestBase: makeRequest(cid),
+		ProcessID:   p.id,
+		TimeoutInMs: 0xffffffff,
+	}
+	p.waitCall, err = gc.brdg.AsyncRPC(ctx, rpcWaitForProcess, &waitReq, &p.waitResp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait on process, leaking process: %s", err)
+	}
+	return p, nil
+}
+
+func (gc *GuestConnection) CreateProcess(settings interface{}) (cow.Process, error) {
+	return gc.exec(context.TODO(), nullContainerID, settings)
+}
+
+func (p *Process) Close() error {
+	p.stdin.Close()
```

Comment:
Why don't you have to check for `nil` here?

Reply:
Because iochannel's Close checks for a nil receiver.

Reply:
Well I clearly don't understand go. I did not know that this worked. I thought a nil ptr to a type would panic at invocation. Not panic in the receiver function if it didn't check for nil.

Reply:
I was confused when I read this. This makes perfect sense